### DISCUSSION TITLE
Extend the library to work with SCTP too

### DISCUSSION THREAD

**Discussion by uzairzia** [libuv/libuv#4087] [Ideas]
I would like to extend the library to allow creation and handling of asynchronous SCTP sockets.

I read the following line in the contributions page:
> "The stable branch is effectively frozen; patches that change the libuv API/ABI or affect the run-time behavior of applications get rejected." 

If this feature is added, then the API (uv.h) will change slightly. i.e., A few methods will be added to it.

How may I proceed as I do not want my changes to get rejected?

**Comment by saghul** [comment]
Would you really need to extend libuv for that tough? I'd assume you could either integrate usrsctp using an uv_poll_t handle or in a similar way if you lift dcsctp off Chromium.

**Reply by uzairzia** [reply]
What if usrsctp is not used? Sockets may be created using Linux API (sys/socket) and used with poll handles. Would do you recommend this?

**Reply by saghul** [reply]
Sure, why not?

**Comment by bnoordhuis** [comment]
The docs could perhaps be more explicit: adding functions is okay (doesn't break ABI) but adding fields to structs isn't.

(Or deleting fields, or reordering fields, etc. Changing function prototypes is off the table for the same reason.)

But I second Saúl's comment, it's probably not necessary to change libuv in the first place.

**Comment by uzairzia** [comment]
The application for which SCTP functionality is required already uses libUV for UDP and TCP. After examining the libUV code for TCP related functionality (`uv_tcp_t`, `uv_tcp_init`, `uv_tcp_bind`...), it seemed very doable to do the same for SCTP. 

Now there is a SCTP client and server set up using libUV. Its methods and flow control is same as that of libUV's TCP. So the important thing to note is that under-the-hood it is treated as a Stream object, just like TCP.

1. What would be your thoughts on such an implementation?

2. Wouldn't such an implementation that uses the already tried and tested libUV code work better than the one that uses poll handles in conjunction with usrsctp keeping in mind that the code for`uv_sctp_t` emulates that of `uv_tcp_t`? From the docs:
> Using uv_poll_t for any other purpose is not recommended; [uv_tcp_t](https://docs.libuv.org/en/v1.x/tcp.html#c.uv_tcp_t), [uv_udp_t](https://docs.libuv.org/en/v1.x/udp.html#c.uv_udp_t), etc. provide an implementation that is faster and more scalable than what can be achieved with [uv_poll_t](https://docs.libuv.org/en/v1.x/poll.html#c.uv_poll_t),

**Comment by uzairzia** [comment]
To give an overview of the changes:

```
[*****@***** libuv]$ git diff --stat c5644368b56872c4044946c35fa91f09adfadce <MY_LOCAL_COMMIT_HASH>
 include/uv.h        |  34 ++++++++++++
 include/uv/unix.h   |   3 ++
 src/unix/core.c     |   5 ++
 src/unix/internal.h |   4 ++
 src/unix/stream.c   |  29 +++++++++--
 src/unix/tcp.c      | 260 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 src/uv-common.c     |  39 ++++++++++++++
 src/uv-common.h     |  11 ++++
 8 files changed, 378 insertions(+), 7 deletions(-)
```

**Comment by saghul** [comment]
Is there any API missing that prevents `uv_sctp_t` being implemented on top of `uv_udp_t` and `uv_tcp_t` at the application level?

**Reply by uzairzia** [reply]
(To the best of my knowledge)

Firstly, Implementations of `uv_tcp_t` and `uv_udp_t` create sockets of their own respective types (i.e., TCP protocol and UDP protocol). So I am not exactly sure how `uv_sctp_t` can be created on top of `uv_tcp_t` / `uv_udp_t`.

Secondly, if it is talked about encapsulating SCTP traffic inside UDP/TCP then there comes the performance and scalability issue because each SCTP packet would be incorporated inside a TCP/UDP packet. This implies double work at the sending and receiving end (format data according to SCTP, then according to TCP/UDP).

**Reply by saghul** [reply]
Sorry, I should have asked first: are you looking for using native SCTP or SCTP over UDP/TCP like WebRTC data channels do?

**Reply by uzairzia** [reply]
Using native SCTP is the preferred choice here due to scalability and performance considerations.

**Comment by uzairzia** [comment]
How about creating a patch for the changes above and applying the patch whenever the version of libUV is upgraded on the deployed system?

In this way the open-source libuv code will not be changed, and the desired functionality at this end would also be achieved. 

Your feedback would be appreciated.

**Reply by saghul** [reply]
Of course you could do that. But if your code is platform dependent then using uv_poll_t and doing your own read / wrote operations will be equivalent and more maintainable because there is no risk in your patch not applying and since the API is stable it will be more future proof.

**Comment by uzairzia** [comment]
FYI:
This thread will be closed on Monday, 17 July 2023 at 12:00 GMT.

**Comment by uzairzia** [comment]
Thank you @saghul and @bnoordhuis for your valuable guidance and time.
